Source/DafnyStandardLibraries/src/DafnyStdLibs/Base64.dfy

```diff
@@ -0,0 +1,1465 @@
+  {
+    reveal BV24ToSeq();
+    reveal SeqToBV24();
+  }
+
+  lemma SeqToBV24ToSeq(s: seq<bv8>)
+    requires |s| == 3
+    ensures BV24ToSeq(SeqToBV24(s)) == s
+  {
+    reveal SeqToBV24();
+    reveal BV24ToSeq();
+  }
+
+  opaque function BV24ToIndexSeq(x: bv24): (ret: seq<index>)
+    ensures |ret| == 4
+  {
+    var b0 := ((x >> 18) & 0x00003F) as index;
+    var b1 := ((x >> 12) & 0x00003F) as index;
+    var b2 := ((x >>  6) & 0x00003F) as index;
+    var b3 := ((x      ) & 0x00003F) as index;
+    [b0, b1, b2, b3]
+  }
+
+  opaque function IndexSeqToBV24(x: seq<index>): (ret: bv24)
+    requires |x| == 4
+  {
+    (x[0] as bv24 << 18) |
+    (x[1] as bv24 << 12) |
+    (x[2] as bv24 <<  6) |
+    (x[3] as bv24      )
+  }
+
+  lemma BV24ToIndexSeqToBV24(x: bv24)
+    ensures IndexSeqToBV24(BV24ToIndexSeq(x)) == x
+  {
+    reveal IndexSeqToBV24();
+    reveal BV24ToIndexSeq();
+  }
+
+  lemma IndexSeqToBV24ToIndexSeq(s: seq<index>)
+    requires |s| == 4
+    ensures BV24ToIndexSeq(IndexSeqToBV24(s)) == s
+  {
+    reveal IndexSeqToBV24();
+    reveal BV24ToIndexSeq();
+  }
+
+  opaque function DecodeBlock(s: seq<index>): (ret: seq<bv8>)
+    requires |s| == 4
+    ensures |ret| == 3
+  {
+    BV24ToSeq(IndexSeqToBV24(s))
+  }
+
+  opaque function EncodeBlock(s: seq<bv8>): (ret: seq<index>)
+    requires |s| == 3
+    ensures |ret| == 4
+  {
+    BV24ToIndexSeq(SeqToBV24(s))
+  }
+
+  lemma EncodeDecodeBlock(s: seq<bv8>)
+    requires |s| == 3
+    ensures DecodeBlock(EncodeBlock(s)) == s
+  {
+    reveal EncodeBlock();
+    reveal DecodeBlock();
+    var b := SeqToBV24(s);
+    BV24ToIndexSeqToBV24(b);
+    SeqToBV24ToSeq(s);
+  }
+
+  lemma DecodeEncodeBlock(s: seq<index>)
+    requires |s| == 4
+    ensures EncodeBlock(DecodeBlock(s)) == s
+  {
+    reveal EncodeBlock();
+    reveal DecodeBlock();
+    var b := IndexSeqToBV24(s);
+    BV24ToSeqToBV24(b);
+    IndexSeqToBV24ToIndexSeq(s);
+  }
+
+  opaque function DecodeRecursively(s: seq<index>): (b: seq<bv8>)
+    requires |s| % 4 == 0
+    decreases |s|
+  {
+    if |s| == 0 then []
+    else DecodeBlock(s[..4]) + DecodeRecursively(s[4..])
+  }
+
+  lemma DecodeRecursivelyBounds(s: seq<index>)
+    requires |s| % 4 == 0
+    ensures |DecodeRecursively(s)| == |s| / 4 * 3
+    ensures |DecodeRecursively(s)| % 3 == 0
+    ensures |DecodeRecursively(s)| == 0 ==> |s| == 0
+  {
+    reveal DecodeRecursively();
+  }
+
+  lemma DecodeRecursivelyBlock(s: seq<index>)
+    requires |s| % 4 == 0
+    ensures
+      (DecodeRecursivelyBounds(s);
+       var b := DecodeRecursively(s);
+       |b| != 0 ==> EncodeBlock(b[..3]) == s[..4])
+  {
+    DecodeRecursivelyBounds(s);
+    if |s| == 0 {}
+    else {
+      DecodeEncodeBlock(s[..4]);
+      reveal DecodeRecursively();
+    }
+  }
+
+  opaque function EncodeRecursively(b: seq<bv8>): (s: seq<index>)
+    requires |b| % 3 == 0
+  {
+    if |b| == 0 then []
+    else EncodeBlock(b[..3]) + EncodeRecursively(b[3..])
+  }
```

Comment:
I see this function is auto-accumulator tail-recursive already, so providing a straightforward method implementation would not do any good.
However, I just proved this potential optimization, could you test the performance of the following, especially for the test that is not working in Go?

```suggestion
  } by method {
    var resultLength := |b|/3 * 4;
    var result := new index[resultLength];
    var i := |b|;
    var j := resultLength;
    
    reveal EncodeRecursively();
    while i > 0
      invariant i % 3 == 0
      invariant 0 <= i <= |b|
      invariant i * 4 == j * 3
      invariant 0 <= j <= resultLength
      invariant result[j..] == EncodeRecursively(b[i..])
    {
      i := i - 3;
      j := j - 4;
      var block := EncodeBlock(b[i..i+3]);
      result[j] := block[0];
      result[j+1] := block[1];
      result[j+2] := block[2];
      result[j+3] := block[3];
      assert result[j+4..] == EncodeRecursively(b[i+3..]);
      assert b[i..][..3] == b[i..i+3];
      assert b[i..][3..] == b[i+3..];
      assert result[j..j+4] == block;
      calc {
        EncodeBlock(b[i..i+3]) + EncodeRecursively(b[i+3..]);
        EncodeBlock(b[i..][..3]) + EncodeRecursively(b[i..][3..]);
        EncodeRecursively(b[i..]);
      }
    }
    s := result[..];
  }
```

Reply:
Hah! I had just been writing this up, myself. I'll take the best of both of our versions and include them in an update.